File: backend/internal/services/audit_service.go
package services

import (
	"encoding/json"
	"tea-distribution/internal/db"
	"tea-distribution/internal/models"

	"github.com/google/uuid"
)

type AuditService struct{}

func NewAuditService() *AuditService {
	return &AuditService{}
}

func (s *AuditService) Log(
	entityType string,
	entityID uuid.UUID,
	action string,
	oldStatus string,
	newStatus string,
	operatorID uuid.UUID,
	operatorName string,
	changes interface{},
	remark string,
	ip string,
	userAgent string,
) error {
	var changesStr string
	if changes != nil {
		changesBytes, _ := json.Marshal(changes)
		changesStr = string(changesBytes)
	}

	log := &models.AuditLog{
		EntityType:   entityType,
		EntityID:     entityID,
		Action:       action,
		OldStatus:    oldStatus,
		NewStatus:    newStatus,
		OperatorID:   operatorID,
		OperatorName: operatorName,
		Changes:      changesStr,
		Remark:       remark,
		IPAddress:    ip,
		UserAgent:    userAgent,
	}

	return db.DB.Create(log).Error
}

func (s *AuditService) LogStatusChange(
	entityType string,
	entityID uuid.UUID,
	oldStatus string,
	newStatus string,
	operatorID uuid.UUID,
	operatorName string,
	remark string,
) error {
	return s.Log(
		entityType,
		entityID,
		"status_change",
		oldStatus,
		newStatus,
		operatorID,
		operatorName,
		nil,
		remark,
		"",
		"",
	)
}

func (s *AuditService) LogCreate(
	entityType string,
	entityID uuid.UUID,
	operatorID uuid.UUID,
	operatorName string,
	data interface{},
) error {
	return s.Log(
		entityType,
		entityID,
		"create",
		"",
		"",
		operatorID,
		operatorName,
		data,
		"创建记录",
		"",
		"",
	)
}

func (s *AuditService) LogUpdate(
	entityType string,
	entityID uuid.UUID,
	operatorID uuid.UUID,
	operatorName string,
	changes interface{},
	remark string,
) error {
	return s.Log(
		entityType,
		entityID,
		"update",
		"",
		"",
		operatorID,
		operatorName,
		changes,
		remark,
		"",
		"",
	)
}

func (s *AuditService) List(entityType string, entityID uuid.UUID, page, pageSize int) ([]models.AuditLog, int64, error) {
	var logs []models.AuditLog
	var total int64

	query := db.DB.Model(&models.AuditLog{}).Preload("Operator")
	if entityType != "" {
		query = query.Where("entity_type = ?", entityType)
	}
	if entityID != uuid.Nil {
		query = query.Where("entity_id = ?", entityID)
	}

	if err := query.Count(&total).Error; err != nil {
		return nil, 0, models.AppErrInternal("查询审计日志失败")
	}

	offset := (page - 1) * pageSize
	if err := query.Order("created_at DESC").Offset(offset).Limit(pageSize).Find(&logs).Error; err != nil {
		return nil, 0, models.AppErrInternal("查询审计日志失败")
	}

	return logs, total, nil
}

File: server/src/services/auditService.js
import prisma from '../utils/prisma.js';

class AuditService {
  async log({ action, entityType, entityId, beforeValue, afterValue, operatorId, ipAddress, userAgent, requestId }) {
    try {
      await prisma.auditLog.create({
        data: {
          action,
          entityType,
          entityId,
          beforeValue: beforeValue ? JSON.stringify(beforeValue) : null,
          afterValue: afterValue ? JSON.stringify(afterValue) : null,
          operatorId,
          ipAddress,
          userAgent,
          requestId,
        },
      });
    } catch (error) {
      console.error('审计日志写入失败:', error);
    }
  }

  async logOrderStatusChange({ orderId, beforeStatus, afterStatus, operatorId, ipAddress, requestId }) {
    await this.log({
      action: 'ORDER_STATUS_CHANGE',
      entityType: 'Order',
      entityId: orderId,
      beforeValue: { status: beforeStatus },
      afterValue: { status: afterStatus },
      operatorId,
      ipAddress,
      requestId,
    });
  }

  async logOrderUpdate({ orderId, beforeValue, afterValue, operatorId, ipAddress, requestId }) {
    await this.log({
      action: 'ORDER_UPDATE',
      entityType: 'Order',
      entityId: orderId,
      beforeValue,
      afterValue,
      operatorId,
      ipAddress,
      requestId,
    });
  }

  async logWasteRecord({ recordId, operatorId, ipAddress, requestId }) {
    await this.log({
      action: 'WASTE_RECORD_CREATE',
      entityType: 'WasteRecord',
      entityId: recordId,
      afterValue: { id: recordId },
      operatorId,
      ipAddress,
      requestId,
    });
  }

  async logRefund({ refundId, action, operatorId, ipAddress, requestId }) {
    await this.log({
      action,
      entityType: 'Refund',
      entityId: refundId,
      afterValue: { id: refundId },
      operatorId,
      ipAddress,
      requestId,
    });
  }

  async getLogs({ entityType, entityId, action, operatorId, startDate, endDate, page = 1, pageSize = 20 }) {
    const where = {};
    
    if (entityType) where.entityType = entityType;
    if (entityId) where.entityId = entityId;
    if (action) where.action = action;
    if (operatorId) where.operatorId = operatorId;
    
    if (startDate || endDate) {
      where.createdAt = {};
      if (startDate) where.createdAt.gte = new Date(startDate);
      if (endDate) where.createdAt.lte = new Date(endDate);
    }

    const [logs, total] = await Promise.all([
      prisma.auditLog.findMany({
        where,
        orderBy: { createdAt: 'desc' },
        skip: (page - 1) * pageSize,
        take: pageSize,
        include: {
          operator: {
            select: { id: true, name: true, role: true },
          },
        },
      }),
      prisma.auditLog.count({ where }),
    ]);

    return {
      logs: logs.map(log => ({
        ...log,
        beforeValue: log.beforeValue ? JSON.parse(log.beforeValue) : null,
        afterValue: log.afterValue ? JSON.parse(log.afterValue) : null,
      })),
      total,
      page,
      pageSize,
    };
  }
}

export default new AuditService();
